gtsam/hybrid/HybridFactorGraph.cpp

```diff
@@ -0,0 +1,425 @@
+ * @brief  Hybrid factor graph that uses type erasure
+ * @author Fan Jiang
+ * @author Varun Agrawal
+ * @author Frank Dellaert
+ * @date   Mar 11, 2022
+ */
+
+#include <gtsam/base/utilities.h>
+#include <gtsam/discrete/Assignment.h>
+#include <gtsam/discrete/DiscreteEliminationTree.h>
+#include <gtsam/discrete/DiscreteFactorGraph.h>
+#include <gtsam/discrete/DiscreteJunctionTree.h>
+#include <gtsam/hybrid/GaussianMixtureConditional.h>
+#include <gtsam/hybrid/GaussianMixtureFactor.h>
+#include <gtsam/hybrid/HybridConditional.h>
+#include <gtsam/hybrid/HybridDiscreteFactor.h>
+#include <gtsam/hybrid/HybridEliminationTree.h>
+#include <gtsam/hybrid/HybridFactor.h>
+#include <gtsam/hybrid/HybridFactorGraph.h>
+#include <gtsam/hybrid/HybridGaussianFactor.h>
+#include <gtsam/hybrid/HybridJunctionTree.h>
+#include <gtsam/inference/EliminateableFactorGraph-inst.h>
+#include <gtsam/inference/Key.h>
+#include <gtsam/linear/GaussianConditional.h>
+#include <gtsam/linear/GaussianEliminationTree.h>
+#include <gtsam/linear/GaussianFactorGraph.h>
+#include <gtsam/linear/GaussianJunctionTree.h>
+#include <gtsam/linear/HessianFactor.h>
+#include <gtsam/linear/JacobianFactor.h>
+
+#include <algorithm>
+#include <cstddef>
+#include <iostream>
+#include <iterator>
+#include <memory>
+#include <stdexcept>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+namespace gtsam {
+
+template class EliminateableFactorGraph<HybridFactorGraph>;
+
+static std::string BLACK_BOLD = "\033[1;30m";
+static std::string RED_BOLD = "\033[1;31m";
+static std::string GREEN = "\033[0;32m";
+static std::string GREEN_BOLD = "\033[1;32m";
+static std::string RESET = "\033[0m";
+
+constexpr bool DEBUG = false;
+
+static GaussianMixtureFactor::Sum &addGaussian(
+    GaussianMixtureFactor::Sum &sum, const GaussianFactor::shared_ptr &factor) {
+  using Y = GaussianFactorGraph;
+  // If the decision tree is not intiialized, then intialize it.
+  if (sum.empty()) {
+    GaussianFactorGraph result;
+    result.push_back(factor);
+    sum = GaussianMixtureFactor::Sum(result);
+
+  } else {
+    auto add = [&factor](const Y &graph) {
+      auto result = graph;
+      result.push_back(factor);
+      return result;
+    };
+    sum = sum.apply(add);
+  }
+  return sum;
+}
+
+/* ************************************************************************ */
+std::pair<HybridConditional::shared_ptr, HybridFactor::shared_ptr>  //
+EliminateHybrid(const HybridFactorGraph &factors, const Ordering &frontalKeys) {
+  // NOTE(fan): Because we are in the Conditional Gaussian regime there are only
+  // a few cases:
+  // continuous variable, we make a GM if there are hybrid factors;
+  // continuous variable, we make a GF if there are no hybrid factors;
+  // discrete variable, no continuous factor is allowed (escapes CG regime), so
+  // we panic, if discrete only we do the discrete elimination.
+
+  // However it is not that simple. During elimination it is possible that the
```

Comment:
Could you revise the comment to be clear about what was actually implemented. It still reads as "speculative", we "could" etc... It confuses me.

Reply:
I made these changes in https://github.com/varunagrawal/gtsam/pull/46 so maybe @ProfFan can also take a look at it?

Reply:
I’d love this important PR to be self-contained, so could the comment be adapted?